README.md
[![Build Status](https://github.com/fengari-lua/fengari/actions/workflows/ci.yaml/badge.svg)](https://github.com/fengari-lua/fengari/actions/workflows/ci.yaml?query=event%3Apush)
[![npm](https://img.shields.io/npm/v/fengari.svg)](https://npmjs.com/package/fengari)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![#fengari on libera.chat](https://img.shields.io/badge/chat-%23fengari-brightgreen)](https://web.libera.chat/?channels=#fengari)
[![Code Quality: Javascript](https://img.shields.io/lgtm/grade/javascript/g/fengari-lua/fengari.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/fengari-lua/fengari/context:javascript)
[![Total Alerts](https://img.shields.io/lgtm/alerts/g/fengari-lua/fengari.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/fengari-lua/fengari/alerts)

<p align="center">
    <img src="https://github.com/fengari-lua/fengari/raw/master/logo.png" alt="Fengari" width="304" height="304">
</p>


# Fengari

🐺 φεγγάρι - The Lua VM written in JS ES6 for Node and the browser

This repository contains the core fengari code (which is a port of the Lua C library) which includes parser, virtual machine and base libraries.
However it is rare to use this repository directly.

- To use fengari in a web browser as easily as you might use JavaScript, see [fengari-web](https://github.com/fengari-lua/fengari-web)
- [fengari-interop](https://github.com/fengari-lua/fengari-interop) is a lua library that makes interoperating with JavaScript objects simple, it is already included in fengari-web.
- For a clone of the `lua` command line tool, but running under node.js, see [fengari-node-cli](https://github.com/fengari-lua/fengari-node-cli)

### The JS API

Once you've loaded fengari, you can use the JS API:

```js
const luaconf  = fengari.luaconf;
const lua      = fengari.lua;
const lauxlib  = fengari.lauxlib;
const lualib   = fengari.lualib;

const L = lauxlib.luaL_newstate();

lualib.luaL_openlibs(L);

lua.lua_pushliteral(L, "hello world!");
```

The JS API is exactly the same as the C API so `fengari.lua` exposes the same constants and functions as `lua.h`, `fengari.lauxlib` the same as `lauxlib.h` and `fengari.lualib` the same as `lualib.h`. If you're unfamiliar with the C API, you can take a look at [the manual](http://www.lua.org/manual/5.3/manual.html#4).


## Semantics

Fengari implements Lua 5.3 semantics and will hopefully follow future Lua releases. If you find any noticeable difference between Fengari and Lua's behaviours, please [report it](https://github.com/fengari-lua/fengari/issues).

### Strings

Lua strings are 8-bits clean and can embed `\0`. Which means that invalid UTF-8/16 strings are valid Lua strings. Lua functions like `string.dump` even use strings as a way of storing binary data.

To address that issue, Fengari uses [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) objects containing the raw bytes to implement lua strings. To push a JS string on the stack you can use `lua_pushliteral` which will convert it to an array of bytes before pushing it. To get a Lua string on the stack as a JS string you can use `lua_tojsstring` which will attempt to convert it to a UTF-16 JS string. The latter won't give you what you expect if the Lua string is not a valid UTF-16 sequence. You can also convert strings with `luastring_of`, `to_luastring`, `to_jsstring` and `to_uristring`.


### Integers

The JS number type is always a double, and hence cannot accurately represent integers with more than 53 bits. As such, we've taken the route of a rarely used define (`LUA_INT_TYPE=LUA_INT_LONG`) in the PUC-Rio sources, where floats are doubles, but integers are 32 bits.


### `require` and `package.loadlib`

In the browser `require` and `package.loadlib` try to find a file by making synchronous XHR requests.

`require` has been extended to allow searchers to yield.


### _Missing_ features

- `lua_gc`/`collectgarbage`: Fengari relies on the JS garbage collector and does not implement its own.
- The following functions are only available in Node:
    - The entire `io` lib
    - `os.remove`
    - `os.rename`
    - `os.tmpname`
    - `os.execute`
- `debug.debug()` doesn't work from web workers due to lack of a method to get synchronous user input
- [Weak tables](http://www.lua.org/manual/5.3/manual.html#2.5.2)
- `__gc` metamethods


### _Differences_

- `package.jspath` instead of `package.cpath`
- `LUA_JSPATH_DEFAULT` instead of `LUA_CPATH_DEFAULT` (and contains .js extensions rather than .so or .dll extensions)
- `lua_tointegerx` and `lua_tonumberx` do not have out-parameters indicating conversion success. Instead, ``false`` is returned when conversion fails.
- `luaL_execresult` takes an extra argument: an error object. The error object should have fields `status`, `signal` and `errno`.
- `luaL_fileresult` takes an extra argument: an error object. The error object should have a field `errno`.


### Configuring

Some luaconf options can be chosen at library load time. Fengari looks for `process.env.FENGARICONF` and if it exists, parses it as a JSON string.


## Extensions

### `dv = lua_todataview(L, idx)`

Equivalent to `lua_tolstring` but returns a [`DataView`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) instead of a string.


### `lua_pushjsfunction(L, func)`

Alias for `lua_pushcfunction`.


### `lua_pushjsclosure(L, func, n)`

Alias for `lua_pushcclosure`.


### `lua_atnativeerror(L, func)`

Sets a function to be called if a native JavaScript error is thrown across a lua pcall.
The function will be run as if it were a message handler (see https://www.lua.org/manual/5.3/manual.html#2.3).
The current message handler will be run after the native error handler returns.


### `b = lua_isproxy(p, L)`

Returns a boolean `b` indicating whether `p` is a proxy (See `lua_toproxy`).
If `L` is non-null, only returns `true` if `p` belongs to the same global state.


### `p = lua_toproxy(L, idx)`

Returns a JavaScript object `p` that holds a reference to the lua value at the stack index `idx`.
This object can be called with a lua_State to push the value onto that state's stack.

This example would be an inefficient way to write `lua_pushvalue(L, 1)`:

```js
var p = lua_toproxy(L, 1);
p(L);
```


### `fengari` library

A library containing metadata about the fengari release.

  - `AUTHORS`
  - `COPYRIGHT`
  - `RELEASE`
  - `VERSION`
  - `VERSION_MAJOR`
  - `VERSION_MINOR`
  - `VERSION_NUM`
  - `VERSION_RELEASE`

This library is automatically loaded by `luaL_openlibs` into the global `"fengari"`.


## NYI

- `io.input()`: partially implemented
- `io.lines()`
- `io.open()`
- `io.output()`: partially implemented
- `io.popen()`
- `io.read()`
- `io.tmpfile()`
- `file:lines()`
- `file:read()`
- `file:setvbuf()`
- `file:__gc()`


## References

- [Source code for Lua 5.3](lua.org/source/5.3/)
- [Lua 5.2 Bytecode and Virtual Machine](http://files.catwell.info/misc/mirror/lua-5.2-bytecode-vm-dirk-laurie/lua52vm.html)
- [Lua 5.3 Bytecode Reference](http://the-ravi-programming-language.readthedocs.io/en/latest/lua_bytecode_reference.html)
- [A No-Frills Introduction to Lua 5.1 VM Instructions](http://luaforge.net/docman/83/98/ANoFrillsIntroToLua51VMInstructions.pdf)


NEWS
UNRELEASED

  - Add os.setlocale that only understands the C locale
  - Fix incorrect length for certain tables
  - Remove luai_apicheck


0.1.4 - 2018-11-28

  - Fix bug where nil values are not removed from tables


0.1.3 - 2018-10-29

  - Fix loading bytecode with empty string constant
  - Fix missing luaopen_base export
  - Use only C locale for string->number conversions


0.1.2 - 2018-04-22

  - More node 6 compatibility fixes
  - Use indirect eval to get reference to global environment
  - Move test suite to jest
  - Fixes for os.time and os.date normalisation
  - Use own strftime code (drops strftime npm package dependency)
  - Fix passing JavaScript strings to luaL_add(l)string


0.1.1 - 2018-04-03

  - Fix node 6 compatibility
  - Removed duplicate LUA_PATH entries


0.1.0 - 2018-03-31

  - Fengari core implements Lua 5.3 in ES6
  - Implements the Lua C API
  - Lua standard library implemented on top of "C" API
  - Works in node.js
  - Works in browsers (including IE11+, Safari 8+, Chrome, Firefox and Edge)
  - Lua test suite ported to JavaScript and passes (excluding items below)
  - Known incompatability: garbage collector cannot be controlled
  - Known incompatability: __gc not respected
  - Known incompatability: __mode not respected
  - Known incompatability: io library not available in browsers
  - Known incompatability: io library unable to open new files
  - Known incompatability: io library read functionality is not implemented
  - Known incompatability: some os library functions are not implemented


src/ltm.js
"use strict";

const defs    = require('./defs.js');
const { lua_assert } = require('./llimits.js');
const lobject = require('./lobject.js');
const ldo     = require('./ldo.js');
const lstate  = require('./lstate.js');
const {
    luaS_bless,
    luaS_new
} = require('./lstring.js');
const ltable  = require('./ltable.js');
const ldebug  = require('./ldebug.js');
const lvm     = require('./lvm.js');
const CT      = defs.constant_types;

const luaT_typenames_ = [
    "no value",
    "nil",
    "boolean",
    "userdata",
    "number",
    "string",
    "table",
    "function",
    "userdata",
    "thread",
    "proto" /* this last case is used for tests only */
].map(e => defs.to_luastring(e));

const ttypename = function(t) {
    return luaT_typenames_[t + 1];
};


/*
* WARNING: if you change the order of this enumeration,
* grep "ORDER TM" and "ORDER OP"
*/
const TMS = {
    TM_INDEX:    0,
    TM_NEWINDEX: 1,
    TM_GC:       2,
    TM_MODE:     3,
    TM_LEN:      4,
    TM_EQ:       5,  /* last tag method with fast access */
    TM_ADD:      6,
    TM_SUB:      7,
    TM_MUL:      8,
    TM_MOD:      9,
    TM_POW:     10,
    TM_DIV:     11,
    TM_IDIV:    12,
    TM_BAND:    13 ,
    TM_BOR:     14,
    TM_BXOR:    15,
    TM_SHL:     16,
    TM_SHR:     17,
    TM_UNM:     18,
    TM_BNOT:    19,
    TM_LT:      20,
    TM_LE:      21,
    TM_CONCAT:  22,
    TM_CALL:    23,
    TM_N:       24  /* number of elements in the enum */
};

const luaT_init = function(L) {
    L.l_G.tmname[TMS.TM_INDEX]    = new luaS_new(L, defs.to_luastring("__index", true));
    L.l_G.tmname[TMS.TM_NEWINDEX] = new luaS_new(L, defs.to_luastring("__newindex", true));
    L.l_G.tmname[TMS.TM_GC]       = new luaS_new(L, defs.to_luastring("__gc", true));
    L.l_G.tmname[TMS.TM_MODE]     = new luaS_new(L, defs.to_luastring("__mode", true));
    L.l_G.tmname[TMS.TM_LEN]      = new luaS_new(L, defs.to_luastring("__len", true));
    L.l_G.tmname[TMS.TM_EQ]       = new luaS_new(L, defs.to_luastring("__eq", true));
    L.l_G.tmname[TMS.TM_ADD]      = new luaS_new(L, defs.to_luastring("__add", true));
    L.l_G.tmname[TMS.TM_SUB]      = new luaS_new(L, defs.to_luastring("__sub", true));
    L.l_G.tmname[TMS.TM_MUL]      = new luaS_new(L, defs.to_luastring("__mul", true));
    L.l_G.tmname[TMS.TM_MOD]      = new luaS_new(L, defs.to_luastring("__mod", true));
    L.l_G.tmname[TMS.TM_POW]      = new luaS_new(L, defs.to_luastring("__pow", true));
    L.l_G.tmname[TMS.TM_DIV]      = new luaS_new(L, defs.to_luastring("__div", true));
    L.l_G.tmname[TMS.TM_IDIV]     = new luaS_new(L, defs.to_luastring("__idiv", true));
    L.l_G.tmname[TMS.TM_BAND]     = new luaS_new(L, defs.to_luastring("__band", true));
    L.l_G.tmname[TMS.TM_BOR]      = new luaS_new(L, defs.to_luastring("__bor", true));
    L.l_G.tmname[TMS.TM_BXOR]     = new luaS_new(L, defs.to_luastring("__bxor", true));
    L.l_G.tmname[TMS.TM_SHL]      = new luaS_new(L, defs.to_luastring("__shl", true));
    L.l_G.tmname[TMS.TM_SHR]      = new luaS_new(L, defs.to_luastring("__shr", true));
    L.l_G.tmname[TMS.TM_UNM]      = new luaS_new(L, defs.to_luastring("__unm", true));
    L.l_G.tmname[TMS.TM_BNOT]     = new luaS_new(L, defs.to_luastring("__bnot", true));
    L.l_G.tmname[TMS.TM_LT]       = new luaS_new(L, defs.to_luastring("__lt", true));
    L.l_G.tmname[TMS.TM_LE]       = new luaS_new(L, defs.to_luastring("__le", true));
    L.l_G.tmname[TMS.TM_CONCAT]   = new luaS_new(L, defs.to_luastring("__concat", true));
    L.l_G.tmname[TMS.TM_CALL]     = new luaS_new(L, defs.to_luastring("__call", true));
};

/*
** Return the name of the type of an object. For tables and userdata
** with metatable, use their '__name' metafield, if present.
*/
const __name = defs.to_luastring('__name', true);
const luaT_objtypename = function(L, o) {
    let mt;
    if ((o.ttistable() && (mt = o.value.metatable) !== null) ||
        (o.ttisfulluserdata() && (mt = o.value.metatable) !== null)) {
        let name = ltable.luaH_getstr(mt, luaS_bless(L, __name));
        if (name.ttisstring())
            return name.svalue();
    }
    return ttypename(o.ttnov());
};

const luaT_callTM = function(L, f, p1, p2, p3, hasres) {
    let func = L.top;

    lobject.pushobj2s(L, f); /* push function (assume EXTRA_STACK) */
    lobject.pushobj2s(L, p1); /* 1st argument */
    lobject.pushobj2s(L, p2); /* 2nd argument */

    if (!hasres)  /* no result? 'p3' is third argument */
        lobject.pushobj2s(L, p3); /* 3rd argument */

    if (L.ci.callstatus & lstate.CIST_LUA)
        ldo.luaD_call(L, func, hasres);
    else
        ldo.luaD_callnoyield(L, func, hasres);

    if (hasres) {  /* if has result, move it to its place */
        let tv = L.stack[L.top-1];
        delete L.stack[--L.top];
        p3.setfrom(tv);
    }
};

const luaT_callbinTM = function(L, p1, p2, res, event) {
    let tm = luaT_gettmbyobj(L, p1, event);
    if (tm.ttisnil())
        tm = luaT_gettmbyobj(L, p2, event);
    if (tm.ttisnil()) return false;
    luaT_callTM(L, tm, p1, p2, res, 1);
    return true;
};

const luaT_trybinTM = function(L, p1, p2, res, event) {
    if (!luaT_callbinTM(L, p1, p2, res, event)) {
        switch (event) {
            case TMS.TM_CONCAT:
                return ldebug.luaG_concaterror(L, p1, p2);
            case TMS.TM_BAND: case TMS.TM_BOR: case TMS.TM_BXOR:
            case TMS.TM_SHL: case TMS.TM_SHR: case TMS.TM_BNOT: {
                let n1 = lvm.tonumber(p1);
                let n2 = lvm.tonumber(p2);
                if (n1 !== false && n2 !== false)
                    return ldebug.luaG_tointerror(L, p1, p2);
                else
                    return ldebug.luaG_opinterror(L, p1, p2, defs.to_luastring("perform bitwise operation on", true));
            }
            default:
                return ldebug.luaG_opinterror(L, p1, p2, defs.to_luastring("perform arithmetic on", true));
        }
    }
};

const luaT_callorderTM = function(L, p1, p2, event) {
    let res = new lobject.TValue();
    if (!luaT_callbinTM(L, p1, p2, res, event))
        return null;
    else
        return !res.l_isfalse();
};

const fasttm = function(l, et, e) {
    return et === null ? null :
        (et.flags & (1 << e)) ? null : luaT_gettm(et, e, l.l_G.tmname[e]);
};

const luaT_gettm = function(events, event, ename) {
    const tm = ltable.luaH_getstr(events, ename);
    lua_assert(event <= TMS.TM_EQ);
    if (tm.ttisnil()) {  /* no tag method? */
        events.flags |= 1<<event;  /* cache this fact */
        return null;
    }
    else return tm;
};

const luaT_gettmbyobj = function(L, o, event) {
    let mt;
    switch(o.ttnov()) {
        case CT.LUA_TTABLE:
        case CT.LUA_TUSERDATA:
            mt = o.value.metatable;
            break;
        default:
            mt = L.l_G.mt[o.ttnov()];
    }

    return mt ? ltable.luaH_getstr(mt, L.l_G.tmname[event]) : lobject.luaO_nilobject;
};

module.exports.fasttm           = fasttm;
module.exports.TMS              = TMS;
module.exports.luaT_callTM      = luaT_callTM;
module.exports.luaT_callbinTM   = luaT_callbinTM;
module.exports.luaT_trybinTM    = luaT_trybinTM;
module.exports.luaT_callorderTM = luaT_callorderTM;
module.exports.luaT_gettm       = luaT_gettm;
module.exports.luaT_gettmbyobj  = luaT_gettmbyobj;
module.exports.luaT_init        = luaT_init;
module.exports.luaT_objtypename = luaT_objtypename;
module.exports.ttypename        = ttypename;
